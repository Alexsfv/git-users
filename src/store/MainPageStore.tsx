import { action, makeObservable, observable, runInAction } from "mobx"
import { GithubResponseType, GithubUserItem } from "../types/types"

export class MainPageStore {
    userValue: string = ''
    users: Array<GithubUserItem> = []
    isLoadedUsers: boolean = false
    errorInputMessage: string = ''

    constructor() {
        makeObservable(this, {
            userValue: observable,
            users: observable,
            isLoadedUsers: observable,
            errorInputMessage: observable,
            getUsersFormHandler: action,
            getUsers: action,
            changeUserValue: action,
            changeError: action,
            setIsLoadedUsers: action,
            setUsers: action,
        })
    }

    getUsersFormHandler: React.FormEventHandler = (event) => {
        event.preventDefault()
        this.getUsers()
    }

    changeUserValue: React.ChangeEventHandler<HTMLInputElement> = (event) => {
        this.userValue = event.target.value
        if (this.errorInputMessage) this.changeError('')
    }

    getUsers = async () => {
        try {
            if (this.userValue.length <= 0) {
                this.changeError('Введите имя пользователя')
            } else {
                const url = `https://api.github.com/search/users?q=${this.userValue}&`
                const response: GithubResponseType = await fetch(url).then(res => res.json())
                const users = response.items
                runInAction(() => {
                    this.setUsers(users)
                    this.setIsLoadedUsers(true)
                })
            }
        } catch (e) {
            console.warn('Get users request error: ', e.message)
        }
    }

    setUsers(data: GithubUserItem[]) { this.users = data }

    setIsLoadedUsers(status: boolean) { this.isLoadedUsers = status }

    changeError(message: string) { this.errorInputMessage = message }
}
