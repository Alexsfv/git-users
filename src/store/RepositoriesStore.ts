import { GithubRepository } from './../types/types';
import { makeObservable, observable, runInAction, action } from 'mobx';

export class RepositoriesStore {
    repos: GithubRepository[] = []
    repos_count: number = 0

    constructor() {
        makeObservable(this, {
            repos: observable,
            repos_count: observable,
            getRepos: action
        })
    }

    getRepos = async (login: string) => {
        try {
            const url = `https://api.github.com/users/${login}/repos`
            const response = await fetch(url)

            if (response.status === 200) {  
                const repos = await response.json()
                runInAction(() => {
                    this.repos = repos
                    this.repos_count = repos.length
                })

            }

        } catch (e) {

        }
    }
}