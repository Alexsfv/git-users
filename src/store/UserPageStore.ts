import { runInAction } from 'mobx';
import { observable } from 'mobx';
import { action } from 'mobx';
import { GithubUserPage } from './../types/types';
import { makeObservable } from 'mobx';
import { getTimeString } from '../types/func/time';

export interface IUserPageStore extends GithubUserPage {
    isLoaded: boolean
    isExists: boolean | null
    errorMessage: string
    getUserData: (login: string) => void
    clearUserPage: () => void
    catchErrorRosponse: (status: number) => void
}

const userPageStore: IUserPageStore = {
    isLoaded: false,
    isExists: null,
    errorMessage: '',
    login: '',
    avatar_url: '',
    name: '',
    location: null,
    email: null,
    public_repos: 0,
    public_gists: 0,
    followers: 0,
    following: 0,
    blog: '',
    bio: null,
    created_at: '',
    updated_at: '',
    async getUserData(login) {
        try {
            const url = `https://api.github.com/users/${login}`
            const response = await fetch(url)
            if (response.status === 200) {
                const userData: GithubUserPage = await response.json()
                runInAction(() => {
                    this.isExists = true
                    this.errorMessage = ''
                    this.login = userData.login
                    this.avatar_url = userData.avatar_url
                    this.name = userData.name
                    this.location = userData.location
                    this.email = userData.email
                    this.public_repos = userData.public_repos
                    this.public_gists = userData.public_gists
                    this.followers = userData.followers
                    this.following = userData.following
                    this.blog = userData.blog
                    this.bio = userData.bio
                    this.created_at = getTimeString(userData.created_at)
                    this.updated_at = getTimeString(userData.updated_at)
                })
            } else {
                this.catchErrorRosponse(response.status)
            }
        } catch (e) {
            console.warn(e.message)
        }
    },
    catchErrorRosponse(status: number) {
            if (status === 404) {
                runInAction(() => {
                    this.isExists = false
                    this.errorMessage = 'Пользователь с таким логином не существует'
                })
            } else if (status === 403) {
                runInAction(() => {
                    this.errorMessage = 'Превышен лимит запросов, попробуйте позже'
                })
            } else {
                runInAction(() => {
                    this.errorMessage = 'Что-то пошло не так'
                })
            }
    },
    clearUserPage() {
        this.isLoaded = false
        this.isExists = false
        this.errorMessage = ''
        this.login = ''
        this.avatar_url = ''
        this.name = ''
        this.location = null
        this.email = null
        this.public_repos = 0
        this.public_gists = 0
        this.followers = 0
        this.following = 0
        this.blog = ''
        this.bio = null
        this.created_at = ''
        this.updated_at = ''
    }
}

const userPageAnnotations = {
    isLoaded: observable,
    isExists: observable,
    errorMessage: observable,
    login: observable,
    avatar_url: observable,
    name: observable,
    location: observable,
    email: observable,
    public_repos: observable,
    public_gists: observable,
    followers: observable,
    following: observable,
    blog: observable,
    bio: observable,
    created_at: observable,
    updated_at: observable,
    getUserData: action.bound,
    clearUserPage: action.bound,
    catchErrorRosponse: action.bound
}


export const UserPageStore = (): IUserPageStore => makeObservable(userPageStore, userPageAnnotations)