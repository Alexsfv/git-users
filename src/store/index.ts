export * from './MainPageStore'
export * from './UserPageStore'
export * from './RepositoriesStore'