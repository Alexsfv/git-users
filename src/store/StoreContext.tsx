import React, { createContext, useContext } from 'react'
import { rootStoreType, StoreContextProps } from '../types/types'
import { MainPageStore, UserPageStore } from './index'
import { RepositoriesStore } from './RepositoriesStore'

const rootStore: rootStoreType = {
    mainPage: new MainPageStore(),
    userPage: UserPageStore(),
    repositories: new RepositoriesStore()
}

const storeContext = createContext<rootStoreType | null>(null)

export const StoreContext: React.FC<StoreContextProps> = ({children}) => {
    return (
        <storeContext.Provider value={rootStore}>
            {children}
        </storeContext.Provider>
    )
}

export const useStore = () => {
    const store = useContext(storeContext)
    if (store === null) {
        throw new Error('You call useContext, but component has not context. Try use context.Provider component')
    }
    return store
}
