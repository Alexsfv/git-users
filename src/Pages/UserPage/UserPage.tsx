import { observer } from 'mobx-react'
import React, { useEffect } from 'react'
import { RouteComponentProps } from 'react-router'
import { Link } from 'react-router-dom'
import UserInfoBlock from '../../Components/UserInfoBlock/UserInfoBlock'
import UserMainInfoBlock from '../../Components/UserInfoBlock/UserMainInfoBlock/UserMainInfoBlock'
import Repositories from '../../Components/Repositories/Repositories'
import { useStore } from '../../store/StoreContext'
import './UserPage.scss'

type MatchParamsRouterProps = {
    userLogin: string
}

interface UserPageProps extends RouteComponentProps<MatchParamsRouterProps> {}

const UserPage: React.FC<UserPageProps> = observer((props) => {

    const { userPage, repositories } = useStore()
    const userLogin = props.match.params.userLogin

    useEffect(() => {
        userPage.getUserData(userLogin)
        repositories.getRepos(userLogin)
        return () => {
            userPage.clearUserPage()
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])

    return (
        <div className="container">
            <Link className="btn btn-primary btn-lg btn-block mt-4 font-weight-bold" to="/">На Главную</Link>
            {
                userPage.isExists === true
                    ? <div className="row d-flex flex-nowrap justify-content-around  align-items-start align-content-startmt-5 pt-3 px-3 my-4 text-white bg-primary rounded">
                        <div className="d-flex col-5 align-self-center flex-wrap align-items-start align-content-start">
        
                            <UserInfoBlock
                                title="Город"
                                data={userPage.location}
                            />

                            <UserInfoBlock
                                title="Email"
                                data={userPage.email}
                            />

                            <UserInfoBlock
                                title="Репозитории"
                                data={userPage.public_repos}
                            />

                            <UserInfoBlock
                                title="Gists"
                                data={userPage.public_gists}
                            />

                            <UserInfoBlock
                                title="Подписчики"
                                data={userPage.followers}
                            />

                            <UserInfoBlock
                                title="Подписан"
                                data={userPage.following}
                            />

                            <UserInfoBlock
                                title="Блог"
                                data={userPage.blog}
                            />

                            <UserInfoBlock
                                title="Статус"
                                data={userPage.bio}
                            />

                            <UserInfoBlock
                                title="Создан"
                                data={userPage.created_at}
                            />

                            <UserInfoBlock
                                title="Обновлен"
                                data={userPage.updated_at}
                            />

                        </div>

                        <UserMainInfoBlock />

                    </div>
                    : null
            }
            {
                userPage.errorMessage
                    ? <div className="d-flex">
                        <p className="w-100 text-center text-danger font-weight-bold" style={{lineHeight: '80vh', fontSize: '26px'}}>{userPage.errorMessage}</p>
                    </div>
                    : null
            }

            <Repositories />
        </div>
    )
})

export default UserPage