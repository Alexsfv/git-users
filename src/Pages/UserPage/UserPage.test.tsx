import { render } from 'enzyme'
import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import { rootStore, StoreContext } from '../../store/StoreContext'
import UserPage from './UserPage'

const routeComponentPropsMock = {
    history: {} as any,
    location: {} as any,
    match: {
        isExact: false,
        params: { userLogin: 'login' },
        path: '',
        url: ''
    },
}

describe("UserPage component", () => {
    let jsx: JSX.Element

    beforeEach(() => {
        jsx = (
            <BrowserRouter>
                <StoreContext>
                    <UserPage {...routeComponentPropsMock} />
                </StoreContext>
            </BrowserRouter>
        )
    })

    afterEach(() => {
        rootStore.userPage.setIsExists(false)
        rootStore.userPage.setErrorMessage('')
    })

    it("render user info", () => {
        rootStore.userPage.setIsExists(true)
        const component = render(jsx)
        expect(component).toMatchSnapshot()
    })

    it("render without data and without errorMessage", () => {
        const component = render(jsx)
        expect(component).toMatchSnapshot()
    })

    it("render with error message", () => {
        rootStore.userPage.setErrorMessage('wrong user')
        const component = render(jsx)
        expect(component).toMatchSnapshot()
    })
})