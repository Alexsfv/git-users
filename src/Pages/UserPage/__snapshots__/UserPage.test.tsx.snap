// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPage component render user info 1`] = `
<div
  class="container"
>
  <a
    class="btn btn-primary btn-lg btn-block mt-4 font-weight-bold"
    href="/"
  >
    На Главную
  </a>
  <div
    class="row d-flex flex-nowrap justify-content-around  align-items-start align-content-startmt-5 pt-3 px-3 my-4 text-white bg-primary rounded"
  >
    <div
      class="d-flex col-5 align-self-center flex-wrap align-items-start align-content-start"
    >
      <div
        class="d-flex w-100 p-3 mb-3 bg-light rounded shadow text-dark font-weight-bold"
      >
        <p
          class="w-100 m-0"
          style="font-size:18px"
        >
          Репозитории: 
        </p>
        <p
          class="w-100 m-0 text-right"
        >
          0
        </p>
      </div>
      <div
        class="d-flex w-100 p-3 mb-3 bg-light rounded shadow text-dark font-weight-bold"
      >
        <p
          class="w-100 m-0"
          style="font-size:18px"
        >
          Gists: 
        </p>
        <p
          class="w-100 m-0 text-right"
        >
          0
        </p>
      </div>
      <div
        class="d-flex w-100 p-3 mb-3 bg-light rounded shadow text-dark font-weight-bold"
      >
        <p
          class="w-100 m-0"
          style="font-size:18px"
        >
          Подписчики: 
        </p>
        <p
          class="w-100 m-0 text-right"
        >
          0
        </p>
      </div>
      <div
        class="d-flex w-100 p-3 mb-3 bg-light rounded shadow text-dark font-weight-bold"
      >
        <p
          class="w-100 m-0"
          style="font-size:18px"
        >
          Подписан: 
        </p>
        <p
          class="w-100 m-0 text-right"
        >
          0
        </p>
      </div>
    </div>
    <div
      class="card col-5 px-0 mb-3"
      style="width:18rem"
    >
      <a
        href="https://github.com/"
      >
        <img
          alt="userAvatar"
          class="card-img-top"
          src=""
        />
      </a>
      <div
        class="card-body px-3 py-2"
      >
        <div
          class="text-center mt-3"
        >
          <a
            class="btn btn-primary"
            href="https://github.com/"
            target="new_blank"
          >
            Смотреть на GitHub
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    class="row py-4 bg-primary rounded"
  >
    <div
      class="container"
    >
      <div
        class="d-flex flex-wrap justify-content-around"
      >
        <p
          class="w-100 font-weight-bold text-white text-center"
          style="font-size:20px"
        >
          Список репозиториев
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`UserPage component render with error message 1`] = `
<div
  class="container"
>
  <a
    class="btn btn-primary btn-lg btn-block mt-4 font-weight-bold"
    href="/"
  >
    На Главную
  </a>
  <div
    class="d-flex"
  >
    <p
      class="w-100 text-center text-danger font-weight-bold"
      style="line-height:80vh;font-size:26px"
    >
      wrong user
    </p>
  </div>
  <div
    class="row py-4 bg-primary rounded"
  >
    <div
      class="container"
    >
      <div
        class="d-flex flex-wrap justify-content-around"
      >
        <p
          class="w-100 font-weight-bold text-white text-center"
          style="font-size:20px"
        >
          Список репозиториев
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`UserPage component render without data and without errorMessage 1`] = `
<div
  class="container"
>
  <a
    class="btn btn-primary btn-lg btn-block mt-4 font-weight-bold"
    href="/"
  >
    На Главную
  </a>
  <div
    class="row py-4 bg-primary rounded"
  >
    <div
      class="container"
    >
      <div
        class="d-flex flex-wrap justify-content-around"
      >
        <p
          class="w-100 font-weight-bold text-white text-center"
          style="font-size:20px"
        >
          Список репозиториев
        </p>
      </div>
    </div>
  </div>
</div>
`;
